packages/rest/src/lib/handlers/SequentialHandler.ts

```diff
@@ -90,23 +123,56 @@ export class SequentialHandler {
 		// Wait for any previous requests to be completed before this one is run
 		await this.#asyncQueue.wait();
 		try {
-			// Wait for any global rate limits to pass before continuing to process requests
-			await this.manager.globalTimeout;
-			// Check if this request handler is currently rate limited
-			if (this.limited) {
+			/*
+			 * After calculations have been done, pre-emptively stop further requests
+			 * Potentially loop until this task can run if e.g. the global rate limit is hit twice
+			 */
+			while (this.limited) {
+				const isGlobal = this.globalLimited;
+				let limit: number;
+				let timeout: number;
+				let delay: Promise<void>;
+
+				if (isGlobal) {
+					// Set RateLimitData based on the globl limit
+					limit = this.manager.options.globalLimit;
+					timeout = this.manager.globalReset + this.manager.options.offset - Date.now();
+					// If this is the first task to reach the global timeout, set the global delay
+					if (!this.manager.globalDelay) {
+						// The global delay function clears the global delay state when it is resolved
+						this.manager.globalDelay = this.globalDelayFor(timeout);
+					}
+					delay = this.manager.globalDelay;
+				} else {
+					// Set RateLimitData based on the route-specific limit
+					limit = this.limit;
+					timeout = this.timeToReset;
+					delay = sleep(timeout);
+				}
 				// Let library users know they have hit a rate limit
 				this.manager.emit(RESTEvents.RateLimited, {
-					timeToReset: this.timeToReset,
-					limit: this.limit,
+					timeToReset: timeout,
+					limit,
 					method: options.method,
 					hash: this.hash,
 					route: routeId.bucketRoute,
 					majorParameter: this.majorParameter,
+					global: isGlobal,
 				});
-				this.debug(`Waiting ${this.timeToReset}ms for rate limit to pass`);
+				if (isGlobal) {
+					this.debug(`We are globally rate limited, blocking all requests for ${timeout}ms`);
+				} else {
+					this.debug(`Waiting ${timeout}ms for rate limit to pass`);
+				}
 				// Wait the remaining time left before the rate limit resets
-				await sleep(this.timeToReset);
+				await delay;
+			}
+			// As the request goes out, update the global usage information
+			if (!this.manager.globalReset || this.manager.globalReset < Date.now()) {
+				this.manager.globalReset = Date.now() + 1000;
```

Comment:
reset every..second?

Reply:
Yes, this handles the 50/s limit (or whatever you set it to, keeping in mind it defaults to 0/unlimited)